### PR TITLE
Add test decorators that make test cases run multiple times

### LINKED ISSUE #85: Run numerical test multiple times
Related #80 

To reduce undesired failure of numerical tests like `optimizers` or `backward_{cpu,gnu}`, we should add some mechanism in which we run targeted test multiple times and assert fails only if targeted one fails at all attempts.

One of possible implementation is to use decorator like:
```
@test_n_times(10, 8):
def test_something(self):
    ....
```

### PR DESCRIPTION
**Pull request by delta2323**
This PR fixes #85 .
This PR implements decorators for test cases that make them run multiple times.

Decorators are as follows
* repeat_with_success_at_least(n, m) : Pass the test case if it succeeded m times among n trials.
* retry(n) : Pass the test case if it succeeded at least once among n trials.
* repeat(n) : Pass the test case if it succeeded all n trials.

### CODE REVIEW

File: chainer/testing/condition.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import functools
+import unittest
+
+import six
+
+
+class QuietTestRunner(object):
+    def run(self, suite):
```
Comment (unnonouno) [inline chainer/testing/condition.py]: Insert a blank line after `class`

File: chainer/testing/condition.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+                    unittest.FunctionTestCase(
+                        lambda: f(*args, **kwargs),
+                        setUp=cls.setUp,
+                        tearDown=cls.tearDown))
+                if QuietTestRunner().run(suite).wasSuccessful():
+                    success_counter += 1
+                if success_counter >= min_success:
+                    cls.assertTrue(True)
+                    return
+            cls.fail()
```
Comment (unnonouno) [inline chainer/testing/condition.py]: Is it possible to show an original error message?

File: chainer/testing/condition.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    def run(self, suite):
+        result = unittest.TestResult()
+        suite(result)
+        return result
+
+
+def repeat_with_success_at_least(times, min_success):
+    assert times >= min_success
+
+    def _repeat_with_success_at_least(f):
```
Comment (unnonouno) [inline chainer/testing/condition.py]: Do you have a test for this function?

**Comment by unnonouno** [comment]
The code looks good. Please write a document about its specification, and make a unit test.

**Comment by delta2323** [comment]
I added documents and unit tests for the decorators.

File: chainer/testing/condition.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    def run(self, suite):
+        result = unittest.TestResult()
+        suite(result)
+        return result
+
+
+def repeat_with_success_at_least(times, min_success):
+    """Decorator for multiple trial of the test case
+
+    Decorated test case is launched multiple times.
```
Comment (unnonouno) [inline chainer/testing/condition.py]: "A decorated test case" ?

File: chainer/testing/condition.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+class QuietTestRunner(object):
+
+    def run(self, suite):
+        result = unittest.TestResult()
+        suite(result)
+        return result
+
+
+def repeat_with_success_at_least(times, min_success):
+    """Decorator for multiple trial of the test case
```
Comment (unnonouno) [inline chainer/testing/condition.py]: Is "." required?

File: tests/testing_tests/test_condition.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        _should_fail(self, f)
+
+    def test_all_trials_fail2(self):
+        f = self._decorate(MockUnitTest.failure_case, 10, 0)
+        _should_pass(self, f)
+
+    def test_all_trials_succeed(self):
+        f = self._decorate(MockUnitTest.success_case, 10, 10)
+        _should_pass(self, f)
+
```
Comment (unnonouno) [inline tests/testing_tests/test_condition.py]: Can you check the count of the trials. I mean `self.assertEqual(self.unit_test.counter, 10)`.

File: chainer/testing/condition.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+    Decorated test case is launched multiple times.
+    The case is judged as passed at least specified number of trials.
+    If the number of successful trials exceeds `min_success`,
+    the remaining trials are skipped.
+
+    Args:
+        times(int): The number of trials
```
Comment (unnonouno) [inline chainer/testing/condition.py]: "." ?

File: chainer/testing/condition.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        suite(result)
+        return result
+
+
+def repeat_with_success_at_least(times, min_success):
+    """Decorator for multiple trial of the test case
+
+    Decorated test case is launched multiple times.
+    The case is judged as passed at least specified number of trials.
+    If the number of successful trials exceeds `min_success`,
```
Comment (unnonouno) [inline chainer/testing/condition.py]: "``" ?

**Comment by unnonouno** [comment]
LGTM except the comments. And, note that tests show no adequate error messages, as they only show "fail".

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**